### PR TITLE
GS/HW: Allow translating targets for depth sources

### LINKED ISSUE #1824: [GSdx hardware bug] Jackie Chan Adventures flickering and improper cel shading effect
**_Description of the issue_**:

* The black lines in the wall seem to flicker frequently in the hardware renderers, the issue isn't reproducible in the software renderer. (**_GS Dump_**: [**download link**](https://www.dropbox.com/s/w2g9tazwb2bi0nn/JackieChanTextureIssue.7z?dl=0))

  - _D3D11 hardware renderer_
D3D11

  - _OpenGL hardware renderer_
OpenGL

  - _OpenGL software renderer_
SW

* The cel shading effect on the trees will turn out a bit weird on the hardware renderer when you're running around controlling the player. ``( Also note the missing shadows, OpenGL doesn't seem to render the shadows properly at certain scenarios like this one )``

  - _OpenGL hardware renderer_
HW_OGL

  - _OpenGL software renderer_
OGL_SW

**_Note_**: 

 * The D3D renderers have way more issues! D3D9 doesn't even render half of the screen and D3D11 has tons of issues along with no rendering of shadows due to lack of depth read support. I'm just highlighting the OpenGL exclusive bugs in this report.
 
 * Test was done with all accuracy options enabled and tried various combinations of hardware hacks to try and fix the issue  but all the efforts turned out to be futile.

### LINKED ISSUE #8313: [BUG]: Depth Issues and Right side of screen problems in 10 Pin: Champions Alley
### Describe the Bug

There are two bugs with the game when running in HW mode.
1. Right side of menu screens is missing in HW mode. If you switch to SW mode and then back to HW the menu appears correctly.
2. In game have depth issues (mostly in right part of screen). This problem doesn't get solved by switching to SW mode and back. If you enable preload frame data the depth issues get fixed somewhat but then the flames disappear.

### Reproduction Steps

Start the game in HW mode and check menu screen and ingame

### Expected Behavior

Menus and ingame should appear correctly like SW mode.

### PCSX2 Revision

v1.7.4174

### Operating System

Windows 10 (64bit)

### If Linux - Specify Distro

_No response_

### CPU

i3-9100f

### GPU

GTX 1060

### GS Settings

_No response_

### Emulation Settings

_No response_

### GS Window Screenshots

10 Pin - Champions Alley_SLES-53150_20230302201850
10 Pin - Champions Alley_SLES-53150_20230302201908
10 Pin - Champions Alley_SLES-53150_20230302202003
10 Pin - Champions Alley_SLES-53150_20230302202709

### Logs & Dumps

[10 Pin - Champions Alley_SLES-53150_20230302201908.gs.zip](https://github.com/PCSX2/pcsx2/files/10872580/10.Pin.-.Champions.Alley_SLES-53150_20230302201908.gs.zip)

### PR DESCRIPTION
**Pull request by stenzek**
### Description of Changes

Fixes broken half screen in Miami Vice and 10 Pin - Champions Alley.

### Rationale behind Changes

Untitled

Closes #8313.
Closes #1824.

### Suggested Testing Steps

Test affected games.

### CODE REVIEW

**Review by JordanTheToaster — APPROVED** [review]

**Review by Mrlinkwii — APPROVED** [review]

File: pcsx2/GS/Renderers/Metal/tfx.metal [hunk 1]
```diff
@@ -563,9 +563,14 @@ struct PSMain
 	float4 sample_depth(float2 st)
 	{
 		float2 uv_f = float2(clamp_wrap_uv_depth(ushort2(st))) * float2(cb.scale_factor.x);
-		ushort2 uv = ushort2(uv_f);
 
+		#if PS_REGION_RECT == 1
```
Comment (lightningterror) [inline pcsx2/GS/Renderers/Metal/tfx.metal:567]: Maybe better to use runtime if check for consistency?

File: bin/resources/shaders/opengl/tfx_fs.glsl [hunk 2]
```diff
@@ -376,9 +376,14 @@ ivec2 clamp_wrap_uv_depth(ivec2 uv)
 vec4 sample_depth(vec2 st)
 {
     vec2 uv_f = vec2(clamp_wrap_uv_depth(ivec2(st))) * vec2(ScaledScaleFactor);
-    ivec2 uv = ivec2(uv_f);
 
+    #if PS_REGION_RECT == 1
+		uv_f = clamp(uv_f + STRange.xy, STRange.xy, STRange.zw);
+	#endif
```
Comment (lightningterror) [inline bin/resources/shaders/opengl/tfx_fs.glsl:382]: Formatting.

**Comment by refractionpcsx2** [comment]
Fixes the half screen/cell shading issue from #1824 but the missing shadow remains.

Master:
image

PR:
image

**Comment by stenzek** [comment]
Shadows fixed.

**Comment by refractionpcsx2** [comment]
Confirmed :) 
image

**Comment by refractionpcsx2** [comment]
[Psychonauts_SLUS-21120_20230412125111.gs.zip](https://github.com/PCSX2/pcsx2/files/11211002/Psychonauts_SLUS-21120_20230412125111.gs.zip)

Here's a multi-frame dump which does show the problem, which is good cos we just bumped the savestate version so regression testing was going to be impossible :P

**Pull request merged**